corehq/apps/app_manager/static/app_manager/js/report-module.js

```diff
@@ -89,7 +89,95 @@ var ReportModule = (function () {
+                }
+                filter.selectedValue.value = ko.observable(filter.selectedValue.value ? filter.selectedValue.value.join("\u001F") : '');
+
+                filter.dynamicFilterName = ko.computed(function () {
+                    return reportId() + '/' + filter.slug;
+                });
+
+                if(filter.choices !== undefined && filter.show_all) {
+                    filter.choices.unshift({value: "_all", display: "Show All"}); // TODO: translate
+                }
+            }
+        });
+
+        this.selectedFilterStructure = ko.computed(function () { // for the chosen report
+            return self.reportFilters[reportId()];
+        });
+
+        this.toJSON = function () {
+            var selectedFilterStructure = self.selectedFilterStructure();
+            var selectedFilterValues = {};
+            for (var i = 0; i < selectedFilterStructure.length; i++) {
+                var filter = selectedFilterStructure[i];
+                if (filter.selectedValue.doc_type()) {
+                    selectedFilterValues[filter.slug] = {};
+                    selectedFilterValues[filter.slug].doc_type = filter.selectedValue.doc_type();
+                    // Depending on doc_type, pull the correct observables' values
+                    var docTypeToField = {
+                        AutoFilter: 'filter_type',
+                        CustomDataAutoFilter: 'custom_data_property',
+                        StaticChoiceFilter: 'select_value',
+                        StaticDatespanFilter: 'date_range'
+                    };
+                    _.each(docTypeToField, function(field, docType) {
+                        if(filter.selectedValue.doc_type() == docType) {
+                            selectedFilterValues[filter.slug][field] = filter.selectedValue[field]();
+                        }
+                    });
+                    if(filter.selectedValue.doc_type() == 'StaticChoiceListFilter') {
+                        selectedFilterValues[filter.slug].value = filter.selectedValue.value().split("\u001F");
+                    }
+                }
+            }
+            return selectedFilterValues;
+        };
+
+        this.addSubscribersToSaveButton = function() {
+            _.each(this.reportFilters, function(filtersInReport) {
+                for (var i = 0; i < filtersInReport.length; i++) {
+                    var filter = filtersInReport[i];
+                    _.each(filter.selectedValue, function(observable) {
+                        observable.subscribe(changeSaveButton);
+                    });
+                }
+            });
+        };
+
+        // TODO - add user-friendly text
+        this.filterDocTypes = [null, 'AutoFilter', 'StaticDatespanFilter', 'CustomDataAutoFilter', 'StaticChoiceListFilter', 'StaticChoiceFilter'];
+        this.autoFilterTypes = ['case_sharing_group', 'location_id', 'username', 'user_id'];
```

Comment:
Why not `_.keys(_filter_type_to_func)`? Is ordering important? Similarly, could you expose `doc_type_to_filter_class` and use it here rather than repeating the list?